flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/ZooKeeperLeaderElectionDriver.java

```diff
@@ -176,168 +135,92 @@ public void stop() throws Exception{
+					if (LOG.isDebugEnabled()) {
+						LOG.debug(
+							"Leader node changed while {} is the leader with session ID {}.",
+							leaderContenderDesc,
+							confirmedLeaderSessionID);
+					}
+
+					if (confirmedLeaderSessionID != null) {
+						ChildData childData = cache.getCurrentData();
 
-						if (confirmedLeaderSessionID != null) {
-							ChildData childData = cache.getCurrentData();
+						if (childData == null) {
+							if (LOG.isDebugEnabled()) {
+								LOG.debug(
+									"Writing leader information into empty node by {}.",
+									leaderContenderDesc);
+							}
+							writeLeaderInformation(confirmedLeaderSessionID, confirmedLeaderAddress);
+						} else {
+							byte[] data = childData.getData();
 
-							if (childData == null) {
+							if (data == null || data.length == 0) {
+								// the data field seems to be empty, rewrite information
 								if (LOG.isDebugEnabled()) {
 									LOG.debug(
-										"Writing leader information into empty node by {}.",
-										leaderContender.getDescription());
+										"Writing leader information into node with empty data field by {}.",
+										leaderContenderDesc);
 								}
-								writeLeaderInformation();
+								writeLeaderInformation(confirmedLeaderSessionID, confirmedLeaderAddress);
 							} else {
-								byte[] data = childData.getData();
+								ByteArrayInputStream bais = new ByteArrayInputStream(data);
+								ObjectInputStream ois = new ObjectInputStream(bais);
 
-								if (data == null || data.length == 0) {
-									// the data field seems to be empty, rewrite information
+								String leaderAddress = ois.readUTF();
+								UUID leaderSessionID = (UUID) ois.readObject();
+
+								if (!leaderAddress.equals(confirmedLeaderAddress) ||
+									(leaderSessionID == null || !leaderSessionID.equals(confirmedLeaderSessionID))) {
+									// the data field does not correspond to the expected leader information
 									if (LOG.isDebugEnabled()) {
 										LOG.debug(
-											"Writing leader information into node with empty data field by {}.",
-											leaderContender.getDescription());
-									}
-									writeLeaderInformation();
-								} else {
-									ByteArrayInputStream bais = new ByteArrayInputStream(data);
-									ObjectInputStream ois = new ObjectInputStream(bais);
-
-									String leaderAddress = ois.readUTF();
-									UUID leaderSessionID = (UUID) ois.readObject();
-
-									if (!leaderAddress.equals(confirmedLeaderAddress) ||
-										(leaderSessionID == null || !leaderSessionID.equals(confirmedLeaderSessionID))) {
-										// the data field does not correspond to the expected leader information
-										if (LOG.isDebugEnabled()) {
-											LOG.debug(
-												"Correcting leader information by {}.",
-												leaderContender.getDescription());
-										}
-										writeLeaderInformation();
+											"Correcting leader information by {}.",
+											leaderContenderDesc);
```

Comment:
I think it would be good to let the driver extract the leader information and then let the service decide how to handle it (e.g. whether it needs to rewrite the leader information or not).